From c66af75b6b71e1ae2015a34cfd217751d34ddb06 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 08 Feb 2021 13:16:40 +0200
Subject: [PATCH] [SW-20154] gaudi2 : erom arcboot changes

The extended rom APIs are used by
arcboot based on the efuse setting.
Each API is a part its own individual
binary which is indexed in the flash
memory using a separate 'image pointer'
binary.

The 3 erom binaries are:
1) Alternate pcie config (erom_pcie.bin)
2) Recovery image (erom_recovery.bin)
3) TPM (erom_tpm.bin)

From the above binaries, 1) and 2) are
loaded based on efuse bit 13 (logic low)
while 3)is loaded based on efuse bit 11
(logic high).

Change-Id: Icab0e5f2e0855b8c15156c73a792703910437815
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 52df15a..9137d5c 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -20,7 +20,8 @@
 	.recovery_flag = 0,
 	.validation_pass = 0,
 	.efuse_read = 0,
-	.curr_img = 0
+	.curr_img = 0,
+	.erom_boot_desc_flag = 0
 };
 
 #define ARCBOOT_BRINGUP_ERROR		-1
@@ -333,6 +334,8 @@
 	return ARCBOOT_BRINGUP_OK;
 }
 
+__attribute__((weak)) void arcboot_rom_handle_tpm(void) {}
+
 static void arcboot_rom_bringup_pcie(void)
 {
 	void (*ppboot)();
@@ -358,6 +361,9 @@
 			}
 		}
 	}
+
+	arcboot_rom_handle_tpm();
+
 	arcboot_rom_status_update(PROGRESS_PPBOOT_START);
 
 	/* cast ppboot location as function and execute it */
diff --git a/common/boot_desc.c b/common/boot_desc.c
new file mode 100644
index 0000000..359bd81
--- /dev/null
+++ b/common/boot_desc.c
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+#include <string.h>
+#include <stdbool.h>
+#include <hl_fw_if.h>
+#include "hl_boot_desc_erom.h"
+#include <common.h>
+#include <arcboot.h>
+#include "boot_desc.h"
+#include "dma.h"
+#include "spi_flash.h"
+
+#define MAX_IMGS		5
+#define BOOT_PTRS_SZ		sizeof(struct hl_fw_boot_ptrs)
+#define IMG_DESC_SZ		sizeof(struct hl_fw_img_desc)
+#define CRC32_PADD		sizeof(uint32_t)
+#define DESC_MAGIC_MASK		0xFFFFFF00
+#define INV_VER1		((uint32_t)0)
+#define INV_VER2		((uint32_t) -1)
+#define INVALID_DESC_ID		2
+static struct hl_fw_boot_ptrs *p_boot_ptrs;
+static struct hl_fw_img_desc *p_img_desc[MAX_IMGS];
+
+static bool boot_desc_img_desc_is_valid(struct hl_fw_img_desc *img_desc)
+{
+	if ((img_desc->magic.val & DESC_MAGIC_MASK) != HL_FW_IMG_DESC_MAGIC)
+		return false;
+
+	if ((img_desc->desc_ver == INV_VER1) ||
+	    (img_desc->desc_ver == INV_VER2))
+		return false;
+
+	if (img_desc->idx >= INVALID_DESC_ID)
+		return false;
+
+	return true;
+}
+
+int boot_desc_init(void)
+{
+	uint32_t i, boot_ptrs_count, img_ptrs_size;
+	uint32_t *p_count = NULL;
+
+	boot_ptrs_count = spi_flash_readl(HL_BOOT_PTS_COUNT_FLASH_OFFSET);
+
+	img_ptrs_size = BOOT_PTRS_SZ + ((sizeof(uint32_t) +
+			IMG_DESC_SZ) * boot_ptrs_count) + CRC32_PADD;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START);
+
+	/* Fetch boot desc using DMA into SP SRAM offset of 4KB */
+	dma_copy_image(img_ptrs_size, HL_EROM_IMG_PTRS_BASE,
+		       HL_SP_SRAM_BASE + (HL_BOOT_DESC_SP_SRAM_OFFSET >> 2),
+		       SPI_DMA_DST_HBW);
+	if (dma_wait_finish() == DMA_CRC_ERROR)
+		return 0;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH);
+
+	p_count = (uint32_t *) HL_BOOT_PTS_COUNT_SRAM_OFFSET;
+	p_boot_ptrs = (struct hl_fw_boot_ptrs *)(HL_BOOT_PTS_SP_SRAM_START);
+
+	if (p_boot_ptrs->magic.val != HL_FW_BOOT_MAGIC)
+		return 0;
+
+	/* Read all descriptors */
+	for (i = 0; i < p_boot_ptrs->count; i++) {
+		p_img_desc[i] = (struct hl_fw_img_desc *)
+				(HL_BOOT_PTS_SP_SRAM_START + p_count[i]);
+		if (!boot_desc_img_desc_is_valid(p_img_desc[i]))
+			return 0;
+	}
+	arcboot_rom_global.erom_boot_desc_flag = 1;
+	return p_boot_ptrs->count;
+}
+
+uint32_t boot_desc_get_img_desc(enum hl_fw_img_type type,
+				struct boot_desc *desc)
+{
+	uint32_t i;
+
+	/* boot descriptors not initialized */
+	if ((p_boot_ptrs == NULL) || (p_img_desc == NULL))
+		return HL_FW_IMG_NOT_FOUND;
+
+	for (i = 0; i < p_boot_ptrs->count; i++) {
+		if (p_img_desc[i]->magic.magic_type.type != type)
+			continue;
+
+		desc->spi_offset = p_img_desc[i]->offset[p_img_desc[i]->idx];
+		desc->img_size = p_img_desc[i]->size[p_img_desc[i]->idx];
+		desc->loadaddr = ((uint64_t)
+			p_img_desc[i]->load_high[p_img_desc[i]->idx] << 32) |
+			p_img_desc[i]->load_low[p_img_desc[i]->idx];
+		desc->desc_ver = p_img_desc[i]->desc_ver;
+		memcpy(desc->chksum, p_img_desc[i]->chksum[p_img_desc[i]->idx],
+		       ARRAY_SIZE(desc->chksum));
+
+		return p_img_desc[i]->idx;
+	}
+	return HL_FW_IMG_NOT_FOUND;
+}
diff --git a/common/boot_desc.h b/common/boot_desc.h
new file mode 100644
index 0000000..dd90f19
--- /dev/null
+++ b/common/boot_desc.h
@@ -0,0 +1,55 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+#ifndef _BOOT_DESC_H_
+#define _BOOT_DESC_H_
+
+#include <hl_fw_if.h>
+#include <hl_boot_desc_erom.h>
+#include <sha512.h>
+
+#define HL_FW_IMG_NOT_FOUND		((uint32_t) -1)
+/* erom image pointer binary placed in flash at offset 0 */
+#define HL_EROM_IMG_PTRS_BASE		(0x00000000)
+#define HL_CRC32_FOOTER			(sizeof(uint32_t))
+/* erom boot descriptors placed in SP SRAM at offset 4K */
+#define HL_BOOT_DESC_SP_SRAM_OFFSET	(4 * 1024)
+#define HL_BOOT_PTS_SP_SRAM_START	(HL_SCRATCHPAD_SRAM_BASE +\
+					HL_BOOT_DESC_SP_SRAM_OFFSET)
+#define HL_BOOT_PTS_COUNT_SRAM_OFFSET	(HL_BOOT_PTS_SP_SRAM_START +\
+					BOOT_PTRS_SZ)
+#define HL_BOOT_PTS_COUNT_FLASH_OFFSET	(HL_EROM_IMG_PTRS_BASE +\
+					(BOOT_PTRS_SZ - sizeof(uint32_t)))
+
+struct boot_desc {
+	uint32_t desc_ver;
+	uint32_t spi_offset;
+	uint32_t img_size;
+	uint64_t loadaddr;
+	uint8_t chksum[SHA384_SUM_LEN];
+};
+
+/*
+ * boot_desc_init() - read and initialize the boot image pointers structures
+ *
+ * Returns: the amount of images available, or 0 - no images available - legacy
+ */
+int boot_desc_init(void);
+
+/*
+ * boot_desc_get_img_desc() - get the struct boot_desc of the requested image
+ *
+ * @type	- the image type to look for
+ * @desc	- pointer to the struct boot_desc for placing the image data
+ *
+ * Returns:
+ * On success: the index of the image to load
+ * On failure: (u32)-1 (no image by the requested type)
+ */
+uint32_t boot_desc_get_img_desc(enum hl_fw_img_type type,
+				struct boot_desc *desc);
+
+#endif /* _BOOT_DESC_H_ */
diff --git a/common/common.h b/common/common.h
index 9745a39..c6b403c 100644
--- a/common/common.h
+++ b/common/common.h
@@ -11,5 +11,6 @@
 void enable_flr_handling_fsm(void);
 uint32_t get_fld_size_shift(void);
 void update_warm_reset_config(void);
+void arcboot_rom_handle_tpm(void);
 
 #endif /* _COMMON_H_ */
diff --git a/common/hl_boot_desc_erom.h b/common/hl_boot_desc_erom.h
new file mode 100644
index 0000000..0a4b817
--- /dev/null
+++ b/common/hl_boot_desc_erom.h
@@ -0,0 +1,103 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2021 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef HL_BOOT_DESC_EROM_H
+#define HL_BOOT_DESC_EROM_H
+
+#include <stdint.h>
+
+/*
+ * !!!! IMPORTANT INFO !!!!
+ * EROM BOOT DESCRIPTORS (448 BYTES) ARE LOADED INTO SP SRAM at offset 4K
+ * HENCE THIS ADDRESS SHOULD NOT BE USED FOR OTHER PURPOSES.
+ * 448 = sizeof(hl_fw_boot_ptrs) +
+ *       (3 * sizeof(uint32_t) +
+ *       (3 * sizeof(hl_fw_img_desc))
+ * where 3 is the number of erom binaries.
+ */
+
+/* FW boot pointers magic:	  N X T 0 */
+#define HL_FW_BOOT_MAGIC	0x4E585400
+
+/* FW image descriptor magic:	  P T R 0 | <hl_fw_img_type> */
+#define HL_FW_IMG_DESC_MAGIC	0x50545200
+
+/* Current version of the image descriptor */
+#define HL_FW_IMG_DESC_CUR_VER	1
+
+#define HL_CHKSUM_SIZE		48
+
+enum hl_fw_img_type {
+	HL_FW_IMG_ROOT = 0,
+	HL_FW_EROM_PCIE = 1,
+	HL_FW_EROM_RECOVERY = 2,
+	HL_FW_EROM_TPM = 3,
+};
+
+struct boot_magic_type {
+	uint8_t type;	/* enum hl_fw_img_type */
+	char name[3];
+} __packed;
+
+struct hl_fw_boot_magic {
+	union {
+		struct boot_magic_type magic_type;
+		uint32_t val;
+	};
+};
+
+/*
+ *      struct hl_fw_boot_ptrs     struct hl_fw_img_desc
+ *           ============            =================
+ *          || Next     ||    +---->|| Magic         ||
+ *          | ---------- |    |     | --------------- |
+ *          || Mod-able ||    |     || Version       ||
+ *          | ---------- |    |     | --------------- |
+ *          || Magic    ||    |     || Index         ||
+ *          | ---------- |    |     | --------------- |
+ *     +--<<|| Count    ||    |     || Offset    [2] ||
+ *     |     ============     |     | --------------- |
+ *     |                      |     || Size      [2] ||
+ *     |                      |     | --------------- |
+ *     |                      |     || Load High [2] ||
+ *     |                      |     | --------------- |
+ *     |                      |     || Load Low  [2] ||
+ *     |                      |     | --------------- |
+ *     |                      |     || SHA256    [2] ||
+ *     |    Following Array   |     | --------------- |
+ *     |     ============     |      =================
+ *     +--->|| Offset   ||>>--+
+ *     |    | ---------- |           =================
+ *     +--->|| Offset   ||>>------->|  ...            |
+ *     |    | ---------- |           =================
+ *     :       :            :               :
+ *     |    | ---------- |           =================
+ *     +--->|| Offset   ||>>------->|  ...            |
+ *           ============            =================
+ *
+ */
+
+struct hl_fw_boot_ptrs {
+	uint32_t next;	/* backward compatibility */
+	uint32_t mod;	/* backward compatibility */
+	struct hl_fw_boot_magic magic;
+	uint32_t count;
+	/* array of offsets by count size must follow w/o gaps */
+};
+
+struct hl_fw_img_desc {
+	struct hl_fw_boot_magic magic;
+	uint32_t desc_ver;	/* descriptor version */
+	uint32_t idx;		/* "next" index inside the offset array */
+	uint32_t offset[2];	/* the offset on the SPI flash */
+	uint32_t size[2];	/* the size of the image to load */
+	uint32_t load_high[2];	/* the high [32-63] bits of the load address */
+	uint32_t load_low[2];	/* the low [0-31] bits of the load address */
+	uint8_t chksum[2][HL_CHKSUM_SIZE];	/* checksum for image */
+};
+
+#endif /* HL_BOOT_DESC_EROM_H */
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index 1703849..3775721 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -13,8 +13,18 @@
 #include <psoc_global_conf_masks.h>
 #include <psoc_reset_conf_masks.h>
 #include <psoc_reset_conf_regs.h>
+#include <sha512.h>
+#include <string.h>
+#include "dma.h"
+#include "spi_flash.h"
+#include "boot_desc.h"
 
 #define ARCBOOT_STATUS_REG_SZ_BITS	32
+#define LOW_DWORD(x)			(x & 0xFFFFFFFF)
+#define ICCM_OFFS_MASK			0x1FFFF
+#define SP_SRAM_OFFS_MASK		0xFFFF
+#define SRAM_PHYS_OFFS_MASK		0x2FFFFFF
+
 
 void arcboot_rom_status_update(enum arcboot_rom_progress progress)
 {
@@ -65,6 +75,202 @@
 	/* TODO: No HW FSM bits in Gaudi2, wait for ORON to comeback */
 }
 
+static int verify_sha(uint32_t data_size, unsigned char *ptr_to_data,
+		      unsigned char *ptr_to_sha)
+{
+	unsigned char actual_hash[SHA384_SUM_LEN];
+
+	sha384_csum(ptr_to_data, data_size, actual_hash, 0);
+
+	return memcmp(ptr_to_sha, actual_hash, SHA384_SUM_LEN);
+}
+
+static int check_boundary(uint64_t addr, uint32_t size, uint64_t endaddr)
+{
+	return ((addr + size) <= endaddr) ? 0 : 1;
+}
+
+/*
+ * check_load_addr_valid updates 'addr' with the 32bit ARC
+ * view point address used to jump to the code and execute.
+ * If address is valid and within bounds 0 is returned,
+ * else returns 1.
+ */
+static int check_load_addr_valid(struct boot_desc *p_desc, uint32_t *addr)
+{
+	uint32_t addr_low = LOW_DWORD(p_desc->loadaddr);
+	int ret;
+
+	switch (p_desc->loadaddr) {
+	case HL_ICCM_BASE ... HL_ICCM_END:
+		ret = check_boundary(p_desc->loadaddr, p_desc->img_size,
+				HL_ICCM_END);
+		*addr = addr_low & ICCM_OFFS_MASK;
+		break;
+	case HL_SP_SRAM_BASE ... HL_SP_SRAM_END:
+		ret = check_boundary(p_desc->loadaddr, p_desc->img_size,
+				HL_SP_SRAM_END);
+		*addr = HL_SCRATCHPAD_SRAM_BASE +
+			(addr_low & SP_SRAM_OFFS_MASK);
+		/*
+		 * Note that the offset of SP SRAM to be provided to perform
+		 * DMA must be in granuality of 4B. ie.
+		 * (right shift offset 0x1000 by 2).
+		 */
+		p_desc->loadaddr = (HL_SP_SRAM_BASE +
+				((addr_low & SP_SRAM_OFFS_MASK) >> 2));
+		break;
+	case HL_SRAM_PHY_BASE ... HL_SRAM_PHY_END:
+		ret = check_boundary(p_desc->loadaddr, p_desc->img_size,
+				HL_SRAM_PHY_END);
+		*addr = HL_SRAM_LOG_BASE +
+			(addr_low & SRAM_PHYS_OFFS_MASK);
+		break;
+	default:
+		ret = 1;
+		break;
+	}
+	return ret;
+}
+
+static int img_desc_precheck(struct boot_desc *p_desc, uint32_t *addr)
+{
+	if (p_desc->img_size == 0)
+		goto img_desc_invalid;
+
+	if (p_desc->spi_offset + p_desc->img_size > FLASH_VALID_MASK)
+		goto img_desc_invalid;
+
+	if (check_load_addr_valid(p_desc, addr))
+		goto img_desc_invalid;
+
+	return 0;
+
+img_desc_invalid:
+	return 1;
+}
+
+static void ext_rom_load_erom_pcie(void)
+{
+	int err;
+	uint32_t idx;
+	struct boot_desc desc;
+	uint32_t load_addr_logical; /* 32 bit ARC view point address */
+	void (*erom_pcie_func)();
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_PCIE);
+
+	idx = boot_desc_get_img_desc(HL_FW_EROM_PCIE, &desc);
+	if (idx == HL_FW_IMG_NOT_FOUND)
+		return;
+
+	err = img_desc_precheck(&desc, &load_addr_logical);
+	if (err)
+		return;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_PCIE_DMA_START);
+
+	dma_copy_image(desc.img_size, desc.spi_offset,
+		       desc.loadaddr, SPI_DMA_DST_HBW);
+	if (dma_wait_finish() == DMA_CRC_ERROR)
+		return;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH);
+
+	/* Verify SHA excluding the CRC byte */
+	err = verify_sha(desc.img_size - HL_CRC32_FOOTER,
+			(unsigned char *)(load_addr_logical),
+			desc.chksum);
+
+	if (!err) {
+		arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_PCIE_START);
+		erom_pcie_func = (void *)(load_addr_logical);
+		erom_pcie_func();
+		arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_PCIE_FINISH);
+	}
+}
+
+static void ext_rom_load_erom_recovery(void)
+{
+	int err;
+	uint32_t idx;
+	struct boot_desc desc;
+	uint32_t load_addr_logical; /* 32 bit ARC view point address */
+	void (*erom_recov_func)();
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_RECOVERY);
+
+	idx = boot_desc_get_img_desc(HL_FW_EROM_RECOVERY, &desc);
+	if (idx == HL_FW_IMG_NOT_FOUND)
+		return;
+
+	err = img_desc_precheck(&desc, &load_addr_logical);
+	if (err)
+		return;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START);
+
+	dma_copy_image(desc.img_size, desc.spi_offset,
+		       desc.loadaddr, SPI_DMA_DST_HBW);
+	if (dma_wait_finish() == DMA_CRC_ERROR)
+		return;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH);
+
+	/* Verify SHA excluding the CRC byte */
+	err = verify_sha(desc.img_size - HL_CRC32_FOOTER,
+			(unsigned char *)(load_addr_logical),
+			desc.chksum);
+
+	if (!err) {
+		arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_RECOVERY_START);
+		erom_recov_func = (void *)(load_addr_logical);
+		erom_recov_func();
+		arcboot_rom_status_update(
+				PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH);
+	}
+}
+
+static void ext_rom_load_erom_tpm(void)
+{
+	int err = 0;
+	uint32_t idx = 0;
+	struct boot_desc desc;
+	uint32_t load_addr_logical; /* 32 bit ARC view point address */
+	void (*erom_tpm_func)();
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_TPM);
+
+	idx = boot_desc_get_img_desc(HL_FW_EROM_TPM, &desc);
+	if (idx == HL_FW_IMG_NOT_FOUND)
+		return;
+
+	err = img_desc_precheck(&desc, &load_addr_logical);
+	if (err)
+		return;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_TPM_DMA_START);
+
+	dma_copy_image(desc.img_size, desc.spi_offset,
+		       desc.loadaddr, SPI_DMA_DST_HBW);
+	if (dma_wait_finish() == DMA_CRC_ERROR)
+		return;
+
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH);
+
+	/* Verify SHA excluding the CRC byte */
+	err = verify_sha(desc.img_size - HL_CRC32_FOOTER,
+			(unsigned char *)(load_addr_logical),
+			desc.chksum);
+
+	if (!err) {
+		arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_TPM_START);
+		erom_tpm_func = (void *)(load_addr_logical);
+		erom_tpm_func();
+		arcboot_rom_status_update(PROGRESS_ARCBOOT_EROM_TPM_FINISH);
+	}
+}
+
 void execute_pcie_recovery(void)
 {
 	/* if eFUSE == ROM_PCIE_Recovery, execute directly */
@@ -74,10 +280,34 @@
 		return;
 	}
 
+	if (!arcboot_rom_global.erom_boot_desc_flag)
+		if (!boot_desc_init()) {
+			arcboot_rom_status_update(
+					PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
+			return;
+		}
+
 	/* else load PCIE binary from eROM to establish the link */
+	ext_rom_load_erom_pcie();
 
 	/* Further, load the recovery bin from eROM */
+	ext_rom_load_erom_recovery();
+}
 
+void arcboot_rom_handle_tpm(void)
+{
+	/* check fuse bit to call the erom function*/
+	if ((arcboot_rom_global.config_fuse &
+	   EFUSE_SYSCFG_ENABLE_TPM_MASK) != 0) {
+		if (!arcboot_rom_global.erom_boot_desc_flag)
+			if (!boot_desc_init()) {
+				arcboot_rom_status_update(
+					PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
+				return;
+			}
+
+		ext_rom_load_erom_tpm();
+	}
 }
 
 void arcboot_rom_set_spi_write_limit(void)
diff --git a/gaudi2_arc_gnu/soc.h b/gaudi2_arc_gnu/soc.h
index 1778b5f..9c3af90 100644
--- a/gaudi2_arc_gnu/soc.h
+++ b/gaudi2_arc_gnu/soc.h
@@ -7,34 +7,46 @@
 #define _SOC_H_
 #include <gaudi2_regs.h>
 
-#define CPU_FREQ	500000000
-#define CPU_FREQ_US	500	/* CPU_FREQ / 1000000 */
+#define CPU_FREQ			500000000
+#define CPU_FREQ_US			500	/* CPU_FREQ / 1000000 */
 
-#define FW_SIZE					65535
-#define SIG_SIZE				384
-#define KEY_SIZE				784
-#define BLOB_START				4
+#define FW_SIZE				65535
+#define SIG_SIZE			384
+#define KEY_SIZE			784
+#define BLOB_START			4
 
-#define SPI_FLASH_BLOCK_SIZE			1024
+#define SPI_FLASH_BLOCK_SIZE		1024
 
 /* 2MB for Gaudi2 */
-#define FLASH_VALID_MASK			0x1FFFFF
+#define FLASH_VALID_MASK		0x1FFFFF
 
-#define HL_SRAM_BASE				0x1000007FF0000000ULL
-#define HL_SP_SRAM_BASE            		0x1000007FF7FE0000ULL
+#define HL_SRAM_BASE			0x1000007FF0000000ULL
+#define HL_SP_SRAM_BASE			0x1000007FF7FE0000ULL
+#define HL_SRAM_PHY_BASE		0x1000FFFFFD000000ULL
 
 /*
  * This is not SDRAM Base, but the base address of General region
  * into which the rest of the components are mapped
  */
-#define HL_SDRAM_BASE				0x20000000
-#define HL_FLASH_OFFSET				0x6000000
-#define HL_ICCM_OFFSET				0x7040000
-#define HL_SBROM_OFFSET				0x7FC0000
-#define HL_FWSRAM_OFFSET			0x7FD0000
-#define HL_SCRATCHPAD_SRAM_OFFSET		0x7FE0000
-#define HL_PCIE_FWSRAM_OFFSET			0x7FF0000
+#define HL_SDRAM_BASE			0x20000000
+#define HL_SRAM_LOG_BASE		0x11000000
 
-#define HL_CFG_BASE				0xF8000000
+#define HL_FLASH_OFFSET			0x6000000
+#define HL_ICCM_OFFSET			0x7040000
+#define HL_SBROM_OFFSET			0x7FC0000
+#define HL_FWSRAM_OFFSET		0x7FD0000
+#define HL_SCRATCHPAD_SRAM_OFFSET	0x7FE0000
+#define HL_PCIE_FWSRAM_OFFSET		0x7FF0000
+#define HL_CFG_BASE			0xF8000000
+#define HL_ICCM_BASE			(HL_SRAM_BASE + HL_ICCM_OFFSET)
+
+#define HL_ICCM_SIZE			0x20000
+#define HL_SP_SRAM_SIZE			0x10000
+#define HL_SRAM_PHY_SIZE		0x2FFFFF
+
+#define HL_ICCM_END			(HL_ICCM_BASE + HL_ICCM_SIZE - 1)
+#define HL_SP_SRAM_END			(HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE - 1)
+#define HL_SRAM_PHY_END			(HL_SRAM_PHY_BASE +\
+					HL_SRAM_PHY_SIZE - 1)
 
 #endif /* _SOC_H_ */
diff --git a/include/arcboot.h b/include/arcboot.h
index 7526f67..de42345 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -54,6 +54,7 @@
 	uint8_t validation_pass;
 	uint8_t efuse_read;
 	uint8_t curr_img;
+	uint8_t erom_boot_desc_flag;
 };
 
 enum arcboot_rom_bringup_proggress {
@@ -94,7 +95,25 @@
 	PROGRESS_BRINGUP_RAM_SEC_SECURITY,
 	PROGRESS_ARCBOOT_RAM_START,
 	PROGRESS_ARCBOOT_RECOVERY,
-	PROGRESS_ARCBOOT_RECOVERY_DMA		/* 30 */
+	PROGRESS_ARCBOOT_RECOVERY_DMA,		/* 30 */
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH,
+	PROGRESS_ARCBOOT_EROM_PCIE,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START,	/* 35 */
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH,
+	PROGRESS_ARCBOOT_EROM_PCIE_START,
+	PROGRESS_ARCBOOT_EROM_PCIE_FINISH,
+	PROGRESS_ARCBOOT_EROM_RECOVERY,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START,/* 40 */
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_START,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH,
+	PROGRESS_ARCBOOT_EROM_TPM,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_START,	/* 45 */
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH,
+	PROGRESS_ARCBOOT_EROM_TPM_START,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH
 };
 
 enum security_location {
